Performance Page, Aircraft Performance and Damage Update

BradMick opened the pull request:
+ Adds PERF page with data from 0 ft to 8000ft pressure altitude and -40 to 40 degrees C.
+ Deletes Anti-Lift script and replaces it with Hatchet RPM Limiter
+ Numerous MPD page fixes and tweaks
+ Updated WCA's to work with new RPM Limiter and to better mirror the aircraft
+ Added memory point called "mainRotorRPM" at (0.0, 0.0, -0.25) to the model

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fn_getInput.sqf:
```diff
@@ -17,10 +17,31 @@ Author:
+private _collectiveVal     = _heli animationSourcePhase "collective";
+private _cyclicFwdAft      = _heli animationSourcePhase "cyclicForward";
+private _cyclicLeftRight   = _heli animationSourcePhase "cyclicAside";
+private _pedalLeftRight    = (inputAction "HeliRudderRight") - (inputAction "HeliRudderLeft");
```
Could we use a rudder animation source? I am guessing because we don't use it properly rn we can't but could we fix it?

BradMick replied:
There is no animation source for helicopter rudder/pedals in ARMA except for AFM.

mattysmith22 replied:
Noted, thank you

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/headers/core.hpp:
```diff
@@ -1,3 +1,11 @@
 #define STABILTOR_MODE_ALWAYSENABLED 2
 #define STABILTOR_MODE_JOYSTICKONLY 1
```
Just realised, we seem to be saying `STABILTOR` rather than `STABILATOR`?

BradMick replied:
Resolved

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/core/fn_coreUpdate.sqf:
```diff
@@ -0,0 +1,55 @@
+//Update the Drivetrain
+[_heli, _deltaTime] call fza_systems_fnc_drivetrainController;
+
+/*
```
Rather than commenting out, please can you delete?

BradMick replied:
Deleted

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/core/fn_coreVariables.sqf:
```diff
@@ -0,0 +1,67 @@
+_heli setVariable ["fza_systems_battTimer",         _battTime];
+//--Buses
+_heli setVariable ["fza_systems_battBusState",      "OFF"];
+_heli setVariable ["fza_systems_ACBusState",        "OFF"];
```
Please can you change the capitisation to `_acBusState` and `_dcBusState` respectively?

BradMick replied:
Complete. Nerd.

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/drivetrain/fn_drivetrainNoseGearbox1.sqf:
```diff
@@ -0,0 +1,19 @@
+/* ----------------------------------------------------------------------------
+Function: fza_systems_fnc_drivetrainNoseGearbox1
```
Why does this exist?

BradMick replied:
Future growth, there's a plan...forgot about it, will implement after merge.

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/drivetrain/fn_drivetrainNoseGearbox2.sqf:
```diff
@@ -0,0 +1,19 @@
+Author:
+	BradMick
+---------------------------------------------------------------------------- */
+params ["_heli"];
```
Why does this exist?

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/core/fn_coreVariables.sqf:
```diff
@@ -0,0 +1,67 @@
+_heli setVariable ["fza_systems_DCBusState",        "OFF"];
+//--Gen 1 and RTRU 1
+_heli setVariable ["fza_systems_gen1State",         "OFF"];
+_heli setVariable ["fza_systems_rtru1State",        "OFF"];
```
For values that are meant to be just "ON" and "OFF", we should not use string variables, but instead a boolean value. For example `["fza_systems_gen1On", true]`. That way you can do

```sqf
if (_heli getVariable "fza_systems_gen1On") then {
```

BradMick replied:
Fixed.

BradMick replied:
You are not my friend. Changed everything over.

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/hydraulics/fn_hydraulicsPriReservoir.sqf:
```diff
@@ -0,0 +1,48 @@
+};
+//Leak
+if (_priReservoirDamage > SYS_HYD_RES_MIN_DMG) then {
+    _priHydLevel_pct = [_priHydLevel_pct, 0.0, (1 / _curLeakTimer) * _deltaTime] call BIS_fnc_lerp;
```
Under testing, does this leak follow an exponential curve or a linear one, and which should it?

BradMick replied:
Whatchu mean?

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/interact/fn_interactBattSwitch.sqf:
```diff
@@ -0,0 +1,40 @@
+		_battSwitchState = "ON";
+		_heli setVariable ["fza_systems_battSwitchState", _battSwitchState, true];
+
+		false;
```
Why should we not register a click in this scenario?

BradMick replied:
Resolved

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/interact/fn_interactBattSwitch.sqf:
```diff
@@ -0,0 +1,40 @@
+		true;
+	};
+	default {
+		true;
```
Why should we register a click in this scenario? Also with the replacement of `"ON", "OFF"` with `true, false`, we won't need a default case

BradMick replied:
Resolved

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/headers/systems.hpp:
```diff
@@ -0,0 +1,29 @@
+#define SYS_APU_DMG_VAL   0.85
+#define SYS_BATT_DMG_VAL  0.85
+#define SYS_GEN_DMG_VAL   0.85
```
What is the intention of `_VAL`, could `_THRESHOLD` be better?

BradMick replied:
Because VAL is shorter to type, THRESH is acceptable I guess...

mattysmith22 reviewed (changes requested):
So far so good. However, there are a large number of changes that I would like to make for consistency and reliability. Please can you look at and implement these?

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/cfgVehicles.hpp:
```diff
@@ -470,235 +471,372 @@ class CfgVehicles
-			class HitHStabilizerL1
-			{	//causes aircraft to shake slightly when damaged
-				armor = 1;
+			class hitengine1	
```
hitEngine1/2 are the default arma names. For the other components, this would also require changes to the hit point names as well in the model. This would not be a quick fix.

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/cfgVehicles.hpp:
```diff
@@ -730,7 +868,7 @@ class CfgVehicles
 				statement="[this] call fza_fnc_weaponguncontrol;";
 			};
-			class pilotdoor_open
+			/*class pilotdoor_open
```
Removed

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/cfgVehicles.hpp:
```diff
@@ -470,235 +471,372 @@ class CfgVehicles
 					volume = "(1 - camPos) * (transmissionDamage factor [0.45, 0.5]) * (rotorSpeed factor [0.2, 0.5])";
 				};
-				class damageAlarmInt
+				/*class damageAlarmInt
```
Removed

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/cfgVehicles.hpp:
```diff
@@ -284,12 +284,13 @@ class CfgVehicles
 					frequency = "0.66 + rotorSpeed / 3";
 					volume = "(1 - camPos) * (transmissionDamage factor [0.45, 0.5]) * (rotorSpeed factor [0.2, 0.5])";
 				};
+				/*
```
Removed

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/CfgRadio.hpp:
```diff
@@ -222,10 +222,10 @@ class CfgRadio
 		title = ;
 	};
-	class fza_ah64_bt_rotorrpm
+	/*class fza_ah64_bt_rotorrpm
```
Removed

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/cfgVehicles.hpp:
```diff
@@ -470,235 +471,372 @@ class CfgVehicles
+				};*/
 			};
 		};
 		class Damage
```
Removed

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/damage/fn_damageSystem.sqf:
```diff
@@ -34,20 +34,20 @@ if (!(alive _heli)) exitwith {
 };
 
-if (_system == "mala vrtule") then {
+/*if (_system == "mala vrtule") then {
```
Removed

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf:
```diff
@@ -115,12 +115,9 @@ if (fza_ah64_enableClickHelper) then {
-private _eng2Lever = _heli getVariable "fza_sfmplus_engPowerLeverState" select 1;
-private _apuState = _heli getVariable "fza_ah64_apu";
-private _powerOnState = _apuState == true || (_eng1state == "ON" && _eng1Lever == "FLY") || (_eng2state == "ON" && _eng2Lever == "FLY");
+private _ACBusState = _heli getVariable "fza_systems_ACBusState";
```
Complete. Nerd.

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/tkoh/fza_ah64d_blockii_exp_orig.xml:
```diff
@@ -1,1758 +0,0 @@
-<?xml version="1.0"?>
-<GenHeli600 version="1.0">
```
Not sure how it made a comeback. I've repeated deleted it.

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fn_getInput.sqf:
```diff
@@ -33,5 +54,54 @@ if (fza_ah64_sfmPlusKeyboardOnly) then {
+};
 
+/*
+[] spawn { 
```
Deleted

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/engine/fn_engineController.sqf:
```diff
@@ -63,18 +70,63 @@ if (isMultiplayer && local _heli && (_heli getVariable "fza_sfmplus_lastTimeProp
+	//systemChat str [_realRPM / 10, _rtrRPM];
+	_rtrRPM = _droopRPM;
+
+	//systemChat str ["adjusted RPM", _rtrRPM];
```
Deleted

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/engine/fn_engineController.sqf:
```diff
@@ -63,18 +70,63 @@ if (isMultiplayer && local _heli && (_heli getVariable "fza_sfmplus_lastTimeProp
+	} else {
+		//systemChat "fixing rotor";
+		_heli setHitpointDamage ["hithrotor", 0.0];
+		_heli engineOn true;
```
When the rotor is damaged to a certain point, it automatically turns the engine off, this is required to keep everything functioning correctly.

mattysmith22 replied:
What I mean is that that we decide whether the engine should be on or not depending on the estarted variable. This doesn't check that before turning it on - it just turns it on no matter what

mattysmith22 replied:
Discussed in a call, we're in agreement this is okay

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/engine/fn_engineController.sqf:
```diff
@@ -63,18 +70,63 @@ if (isMultiplayer && local _heli && (_heli getVariable "fza_sfmplus_lastTimeProp
+if (_heli getHitPointDamage "hithrotor" == 1.0) exitWith {};
+
+private _lastUpdate = _heli getVariable ["fza_sfmplus_lastUpdate", 0];
+if (cba_missionTime > _lastUpdate + 0.3 && _rtrRPM > 0.05) then {
```
Resolved.

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/engine/fn_engine.sqf:
```diff
@@ -118,20 +120,21 @@ private _engBaseTGT      = _intEngBaseTable select 1;
+private _maxTQ_DE   = _heli getVariable "fza_sfmplus_maxTQ_DE";
+private _maxTQ_SE   = _heli getVariable "fza_sfmplus_maxTQ_SE";
+private _maxTQ      = getNumber (_config >> "engMaxTQ");
+/*if (_isSingleEng) then {
```
Deleted

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_mpd/pages/tsd/common.hpp:
```diff
@@ -1,4 +1,6 @@
 #include "components\vabs_top.hpp"
+#include "\fza_ah64_controls\headers\systemConstants.h"
```
Probably a copy pasta error, got lost on my way to somewhere else...

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_mpd/headers/mfdConstants.h:
```diff
@@ -78,8 +78,12 @@
 #define MFD_TEXT_IND_ENG_WCA_6 16
-#define MFD_TEXT_IND_ENG_WCA_7 17
-#define MFD_TEXT_IND_ENG_WCA_8 18
+//#define MFD_TEXT_IND_ENG_WCA_7 17
```
Deleted

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_mpd/headers/mfdConstants.h:
```diff
@@ -57,8 +57,8 @@
 #define MFD_IND_ENG_WCA_6 16
-#define MFD_IND_ENG_WCA_7 17
-#define MFD_IND_ENG_WCA_8 18
+//#define MFD_IND_ENG_WCA_7 17
```
Deleted

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_mpd/functions/fn_update.sqf:
```diff
@@ -21,10 +21,15 @@ Author:
 
-private _mpdsHavePower = (_heli animationphase "plt_apu" > 0.5) || (_heli animationphase "plt_eng1_throttle" >= 0.5) || (_heli animationphase "plt_eng2_throttle" >= 0.5);
+private _mpdsHavePower = false;
+if (_ACBusState == "ON") then {
```
Resolved. Still not friends.

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/damage/fn_damageApply.sqf:
```diff
@@ -22,93 +22,40 @@ params ["_heli", "_deltaTime"];
-}
\ No newline at end of file
+if (_droopRotor) then {
+    fza_sfmplus_liftLossTimer = (fza_sfmplus_liftLossTimer + 0.01) max 0 min 1;
```
Ask Yax

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/engine/fn_engineController.sqf:
```diff
@@ -63,18 +71,58 @@ if (isMultiplayer && local _heli && (_heli getVariable "fza_sfmplus_lastTimeProp
+	if ((_realRPM / 10)  > _rtrRPM) then {
+		_heli setHitpointDamage ["hithrotor", 0.9];
+	} else {
+		_heli setHitpointDamage ["hithrotor", 0.0];
```
Does this line cause your rotor to be repaired if you droop and un-droop it?

mattysmith22 replied:
Discussed in a call, we're in agreement this is okay

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/apu/fn_apu.sqf:
```diff
@@ -0,0 +1,53 @@
+private _apuRPM_pct    = _heli getVariable "fza_systems_apuRPM_pct";
+private _apuFF_kgs     = 0.0;
+
+if (_apuBtnOn && _battBusOn) then {
```
Eventually I would like this to not be a perframe function, but work on timers and eventt handlers. However for now this is okay

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/core/fn_coreVariables.sqf:
```diff
@@ -0,0 +1,68 @@
+//--Buses
+_heli setVariable ["fza_systems_battBusOn",         false];
+_heli setVariable ["fza_systems_acBusOn",           false];
+_heli setVariable ["fza_systems_dcBusOn",           false];
```
These not having a ,true means that they will not be synchronised across multiplayer. Are you sure that none of these need to be synchronised?

If any do, then having them be set in a loop will become untenable - that will cause too much of a network issue

BradMick replied:
Oversight, fixed, they should definitely be synchronized.

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/interact/fn_interactAPUButton.sqf:
```diff
@@ -0,0 +1,47 @@
+_heli setVariable ["fza_systems_apuBtnOn", _apuBtnOn, true];
+
+/*
+switch (_apuBtnOn) do {
```
Please could you remove this commented out code?

BradMick replied:
Fixed.

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/interact/fn_interactAPUButton.sqf:
```diff
@@ -0,0 +1,47 @@
+---------------------------------------------------------------------------- */
+params ["_heli"];
+
+private _apuBtnOn = _heli getVariable "fza_systems_apuBtnOn";
```
This can be swapped into

```sqf
_heli setVariable ["fza_systems_apuBtnOn", !(_heli getVariable "fza_systems_apuBtnOn") true];
```

BradMick replied:
Okie doke. Reads like garbage, but we'll go with that.

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_systems/functions/interact/fn_interactBattSwitch.sqf:
```diff
@@ -0,0 +1,46 @@
+};
+_heli setVariable ["fza_systems_battSwitchOn", _battSwitchOn, true];
+/*
+switch (_battSwitchOn) do {
```
Please can you remove the commented out code?

BradMick replied:
Fixed.

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_ufd/functions/fn_update.sqf:
```diff
@@ -78,10 +81,10 @@ do {
-                if (count _cautions > _i) then {_cautions # _i} else {"           "},
-                if (count _advisories > _i) then {_advisories # _i} else {"          "}
-            ]];
+                if (count _warnings > _i) then {_warnings # _i} else {"          "},       //10
```
The number doesn't matter here, and will lead to inaccuracies as we add / remove more

BradMick replied:
Which number doesn't matter? The number of spaces matters, and keeping track of how many spaces are available for each section does as well.

mattysmith22 reviewed (changes requested):
Looks good, thank you for making the mentioned changes. I have noticed a few other things, commented

BradMick commented on @AH-64D Apache Longbow/Addons/fza_ah64_mpd/pages/eng_draw.hpp:
```diff
@@ -267,8 +273,8 @@ class AirFormatOnly { //<-- This box should ONLY display when there are active w
     MPD_WCA_ITEM(MFD_IND_ENG_WCA_6, MPD_WCA_COLUMN_2_X, 0.6, 1)
-    MPD_WCA_ITEM(MFD_IND_ENG_WCA_7, MPD_WCA_COLUMN_2_X, 0.6, 2)
-    MPD_WCA_ITEM(MFD_IND_ENG_WCA_8, MPD_WCA_COLUMN_2_X, 0.6, 3)
+    //MPD_WCA_ITEM(MFD_IND_ENG_WCA_7, MPD_WCA_COLUMN_2_X, 0.6, 2)
```
Complete

mattysmith22 commented on @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/cfgVehicles.hpp:
```diff
@@ -730,54 +842,6 @@ class CfgVehicles
 				shortcut="OpticsMode";
 				statement="[this] call fza_fnc_weaponguncontrol;";
 			};
-			class pilotdoor_open
```
We could add a ticket to make this an ACE interact so you can do it externally?

mattysmith22 reviewed (approved):
Looks great! Thank you for the help

BradMick commented:
+ Added British Mission Equipment
+ Complete rebuild of the damage/hit point system
+ Added electrical and hydraulic systems and failures
+ Added performance page and updated tables from -40 to 40 deg C and 0 to 8000ft PA
+ Implemented Hatchet rotor RPM limiter
+ Updated WCA page displays
+ Updated audio system for new WCA's and to prevent constant rotor RPM low annunciations
+ Refreshed textures for all aircraft
+ Added custom tail numbers (settable in the 3DEN editor attributes)
+ Added Indian Air Force Livery
+ Added sensorPos memory point for future FCR update
+ Changed initial rotor positioning to match actual aircraft rotor positioning when at rest
